pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -0,0 +1,357 @@
+					id := dns.Id()
+					if _, duplicate = waitingResponses[id]; !duplicate {
+						req.msg.Id = id
+						break
+					}
+				}
+				if duplicate {
+					req.ch <- sharedClientResponse{nil, 0, fmt.Errorf("duplicate request id %d", req.msg.Id)}
+					close(req.ch)
+					continue
+				}
+			}
+
+			err := client.SendContext(req.ctx, req.msg, conn, start)
+			if err != nil {
+				req.ch <- sharedClientResponse{nil, 0, err}
+				close(req.ch)
+			} else {
+				waitingResponses[req.msg.Id] = waiter{req.ch, start}
+
+				// Wake up the receiver that may be waiting to receive again
+				triggerReceiver()
+			}
+
+		case resp, ok := <-responses:
+			if !ok {
+				// 'responses' is closed when the receive loop exits, so we quit as
+				// nothing can be received any more
+				return
+			}
+			if resp.err != nil {
+				// ReadMsg failed, but we cannot match it to a request,
+				// so complete all pending requests.
+				for _, waiter := range waitingResponses {
+					waiter.ch <- sharedClientResponse{nil, 0, resp.err}
+					close(waiter.ch)
+				}
+				waitingResponses = make(map[uint16]waiter)
+			} else if resp.msg != nil {
+				if waiter, ok := waitingResponses[resp.msg.Id]; ok {
+					delete(waitingResponses, resp.msg.Id)
+					resp.rtt = time.Since(waiter.start)
+					waiter.ch <- resp
+					close(waiter.ch)
+				}
+			}
+		}
+	}
+}
+
+func (c *SharedClient) ExchangeSharedContext(ctx context.Context, m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	c.Lock()
+	if c.conn == nil {
+		c.conn, err = c.DialContext(ctx, c.serverAddr)
+		if err != nil {
+			c.Unlock()
+			return nil, 0, fmt.Errorf("failed to dial connection to %v: %w", c.serverAddr, err)
+		}
+		// Start handler for sending and receiving.
+		c.wg.Add(1)
+		go handler(&c.wg, c.Client, c.conn, c.requests)
+	}
+	c.Unlock()
+
+	// This request keeps 'c.requests' open; sending a request may hang indefinitely if
+	// the handler happens to quit at the same time. Use ctx.Done to avoid this.
+	timeout := getTimeoutForRequest(c.Client)
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	respCh := make(chan sharedClientResponse)
+	select {
+	case c.requests <- request{ctx: ctx, msg: m, ch: respCh}:
+	case <-ctx.Done():
+		return nil, 0, ctx.Err()
+	}
+
+	// Since c.requests is unbuffered, the handler is guaranteed to eventually close 'respCh'
+	select {
+	case resp := <-respCh:
+		return resp.msg, resp.rtt, resp.err
+	// This is just fail-safe mechanism in case there is another similar issue
+	case <-time.After(time.Minute):
+		return nil, 0, fmt.Errorf("timeout waiting for response")
+	}
+}
+
+// close closes and waits for the close to finish.
+// Must be called while holding client's lock.
+func (c *SharedClient) close() {
+	close(c.requests)
+	c.wg.Wait()
+	c.conn = nil
+}
+
+// Return the appropriate timeout for a specific request
+func getTimeoutForRequest(c *dns.Client) time.Duration {
+	wtimeout := c.WriteTimeout
+	if wtimeout == 0 {
+		wtimeout = time.Second * 2
```

Comment:
Why did you choose 2 seconds here?

Reply:
I didn't, this was the default in miek/dns 🙃  Added a comment to that effect.